Refactor SqlLifecycle into statement classes

paul-rogers opened the pull request:
With the recent changes to make a single pass through the planner, and to tidy up the JDBC semantics, we are now in a position to tighten up the semantics for SQL statement handling by replacing `SqlLifecycle` with a set of focused “statement” classes (in the sense of following JDBC statements.)

## Motivation

`SqlLifecycle` appears to have been modeled on the original native `QueryLifecycle` which represents native query. While the `QueryLifecycle` works for native queries, SQL imposes additional constraints that suggests a cleaner way to model SQL queries.

In the recent discussion around the “single-pass planner”, @clintropolis  and @gianm provided a helpful description of some of the history as to how the code came to be as it is. One of the major drivers, we discovered, was the need to handle dynamic parameters for JDBC. We discovered that the JDBC implementation made some awkward assumptions, which were fixed in the recent “Avatica” PR.

With the planner and JDBC tidied up, we are now in a position to examine `SqlLifecycle`. This class handles two distinct use cases: the "direct" case (like a JDBC `Statement`) where we execute a given SQL statement, and a "prepared" case (like JDBC `PreparedStatement`) in which we prepare a statement, then execute that statement later with parameters. With the Avatica driver cleaned up to clearly separate these two use cases, we can do the same at the `SqlLifecycle` level.

Since each use case has its own class, the interfaces for the classes are simpler, either:

`directStmt(sql, parameters) -> results`

Or:

`prepare(stmt) -> prepared_stmt`
`prepared_stmt(parameters) -> results`

In the "direct" statements, the query is "one and done". But, with prepared statements (for JDBC), the prepared statement can be run any number of times, each time producing a direct statement with parameters applied.

### Specific Use Cases

For HTTP, we do “direct” statement execution: we are presented with the query, context and parameters, and we make a straight shot through the planner to execution, stopping only if we encounter an error. This is the same pattern (but with somewhat different inputs) used by tests.

On the other hand, for JDBC, we are obligated to implement a “prepare once, execute many” model. https://docs.oracle.com/javase/tutorial/jdbc/basics/prepared.html In our case, this means doing validation without parameter values (which are not yet provided by JDBC.) Then, to execute, we must replan with parameter values. Since we use a “query optimized” model, this means starting over as in the original code.

## Background

In traditional, non-distributed systems, the high-order view of a SQL system is that it is just a function:

```text
run(sql, context) --> results
```

Where `context` is an ad-hoc collection of data required to run a query.

In a distributed system, we often split the above into two steps:

```text
plan(sql, context) --> plan
exec(plan, context) --> results
```

The functions do have internal structure, but that is generally not important to the caller. The existing `SqlLifecycle` exposes that internal structure at each point where the caller must provide more information (such as parameter values, an auth result, etc.) If we instead take that information to be part of the `context` above, then we can simplify the interface to follow the traditional model. (Druid is distributed, but the distribution happens as a level below the SQL statement.)

When a system supports the `PREPARE` operation, then use the second pattern above: we create an execution plan during `PREPARE` and we run the plan, with parameters, as a later step. When a system supports `EXPLAIN`, then we do only the `plan` step above, and skip the execution part. (In practice, the "prepare result" in Druid is not actual an execution plan, but it is close enough for our purposes.)

## Restructured Classes

`SqlLifecycle` is replaced with classes focused on the two use cases identified above: a “direct” statement and a “prepared” statement:

* `DirectStatement`: One which implements the single-pass model, using the auth result for authentication.
* `HttpStatement`: a variation of `DirectStatement` which uses the servlet request for authentication.
* `PreparedStatement`: implements the prepare-once, execute-many model.

The resulting code is far simpler and easier to test. We now have a clean SQL stack for JDBC, HTTP and internal use that make a single planner pass when possible, and use two passes for the JDBC prepare/execute model.

The JDBC classes are then implemented using the statement abstractions described above. A JDBC `ResultSet` is a `DirectStatement`. A JDBC `PreparedStatement` is a Druid `PreparedStatement`. The prior JDBC-related PR did the restructuring that allows us to now map JDBC directly to Druid statements.

## Secondary Changes

Other supporting changes in this PR include:

* The `DruidPlanner` no longer has the “reset” functionality: each planner instance is used by a single statement, only once.
* The code to gather and report stats is factored out into a new `SqlExecutionReporter` class which is shared by the new statement classes.
* The class to cancel queries, `SqlLifecycleManager` is refactored to use a `Cancellable` interface instead of depending on the details of `SqlLifecycle`.
* `SqlToolbox` is added to hold the many knick-knacks passed from the `SqlLivecycleFactory` into statements.
* The old mock-based `SqlLifecycleTest` is replace with a new `SqlStatementTest`.
* `BaseCalciteQueryTest` is modified to use the `SqlQueryPlus` object introduced recently.
* Many tests were updated to remove a now-unnecessary `throws` clause to keep IntelliJ inspections happy.
* `PlannerConfig` us used with many variations in tests. Because there is no constructor, each variation created a subclass to change one value. A builder is added to make customizing the config easier in tests.
* Initialization of the Calcite default character set has moved from code in `Calcites` into a `saffron.properties` file in the `druid-sql` project resources directory. This change avoids any possible race condition between Druid and Calcite initialization: Calcite is guaranteed to obtain properties when it initializes. The properties file is compiled into the `druid-sql` jar file and is thus guaranteed to be available in production.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] been tested in a test Druid cluster.

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:
```diff
@@ -0,0 +1,203 @@
+import java.util.function.Function;
+
+/**
+ * Represents a SQL statement either for preparation or execution.
```
> Closing the statement emits logs and metrics for the statement.

Is there any case when we want to emit logs/metrics beforehand, for example, an exception while calling the public methods of the class? If so, is the above statement a recommendation for implementations or a binding contract?

paul-rogers replied:
Great question. By "beforehand", do you mean "before we close the statement"? The example was "an exception while calling the public methods of the class." Let's start by defining the protocol for the new code. An instance is created with the query info. Then, the query is prepared or run, which can fail. If the calls fail, we pass the exception to the reporter class. Regardless of success or failure, the caller closes the statement. On close, the reporter will emit either the success or failure case. The only way to avoid reporting is to call `closeQuietly()`, as in tests.

Judging by the existing code, it seems we always emit logs and and metrics at the conclusion of the lifecycle: there is a pile of code in `SqlResource` to do that, as well as code in the JDBC classes. The JDBC tests actually verify that the proper log info is produced. In that context, we emit logs and metrics for both successful and failed queries. There was some ambiguity about failures due to a security violation: that info was _not_ logged previously, but a comment suggested this was an undesired limitation rather than a feature. This PR adds logging even for security failures.

To recap, the result is that, if `close()` is called, logs and metrics are emitted for both successful and failed queries. For tests, we call `closeQuietly()` since there is no need to emit this information.

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/SqlQueryPlus.java:
```diff
@@ -66,6 +66,25 @@ public SqlQueryPlus(final String sql, final AuthenticationResult authResult)
     this(sql, (QueryContext) null, null, authResult);
   }
 
+  public SqlQueryPlus(
```
nit: Should we have a Builder for this class considering that there are multiple constructors for it filling in `null` for different parameters?

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/DirectStatement.java:
```diff
@@ -0,0 +1,179 @@
+ * thread creates this object and executes the query. In the split model,
+ * a request thread creates this object and plans the query. A separate
+ * response thread consumes results and performs any desired logging, etc.
+ * The object is transferred between threads, with no overlapping access.
```
>  The object is transferred between threads, with no overlapping access.

Newbie question: Does this mean that only a single thread holds ownership of an object of this class at a time? If so, then `synchronized` methods and locks are not required.

Also, if the object can be seen from different threads, then why isn't `prepareResult` and `plannerResult` declared volatile?

LakshSingla replied:
Thanks for the explanation. IMO there shouldn't be corner cases with this "sequential" manner in which Statements are accessed.

LakshSingla commented on sql/src/main/java/org/apache/druid/sql/SqlExecutionReporter.java:
```diff
@@ -0,0 +1,146 @@
+
+        if (e instanceof QueryInterruptedException || e instanceof QueryTimeoutException) {
+          statsMap.put("interrupted", true);
+          statsMap.put("reason", e.toString());
```
I am unfamiliar with reporting the metrics, but is there an advantage of duplicating `exception` as `reason` in case the query is interrupted? Also, in case the query is not interrupted, then should we have `statsMap.put("interrupted", false);` for symmetry?

paul-rogers replied:
To be honest, I didn't pay too much attention to this code. If you were to diff this function against the original copy, you'd see that only the field names changed: the rest of the code is pretty much how I found it.

That said, yes, it does appear that this code evolved over time and is perhaps not the most optimally structured.

BTW: this is one big hole in the otherwise great GitHub diff mechanism: no easy way to compare code when it moves from one file to another.

gianm commented on sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:
```diff
@@ -0,0 +1,203 @@
+      throw new ForbiddenException(authorizationResult.toString());
+    }
+
+    queryResource = planner.resourceActions(false);
```
Why do we need to pass an explicit `false` here, instead of `authorizeContextParams`? Seems sketchy, so there should be a comment explaining this.

I also suggest having the names be more aligned, since they're very similar things. So, instead of `resourceActions` and `queryResource`, something like `resourceActions` and `resourceActionsWithoutContext`.

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java:
```diff
@@ -244,7 +237,13 @@ public PrepareResult prepare() throws ValidationException
+    return prepareResult;
+  }
 
+  private void doPrepare(QueryMaker queryMaker) throws ValidationException
```
Since `queryMaker` here is nullable, it's nice to include `@Nullable`.

gianm commented on sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:
```diff
@@ -0,0 +1,203 @@
+  public void close()
+  {
+    closeQuietly();
+    reporter.emit();
```
I'm wondering if there should be a try/finally here. Can `closeQuietly` throw errors? If so, should we still call `reporter.emit()` in that case?

gianm commented on sql/src/main/java/org/apache/druid/sql/DirectStatement.java:
```diff
@@ -0,0 +1,179 @@
+
+  protected PrepareResult prepareResult;
+  protected PlannerResult plannerResult;
+  private volatile boolean cancelled;
```
Canceled (spelling (sorry))

gianm commented on sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:
```diff
@@ -0,0 +1,203 @@
+  )
+  {
+    this.sqlToolbox = sqlToolbox;
+    this.queryPlus = sqlRequest;
```
Any reason to use different variable names for the constructor param and the field?

gianm commented on sql/src/main/java/org/apache/druid/sql/PreparedStatement.java:
```diff
@@ -0,0 +1,106 @@
+        queryPlus.sql(),
+        queryPlus.context())) {
+      validate(planner);
+      authorize(planner, authorizer());
```
Shall we re-authorize the statement every time it runs? That way, if someone's permissions are adjusted such that they can no longer run the query, they don't have a backdoor way to keep running it.

gianm replied:
Nevermind, I see that we're already doing that because we return a new DirectStatement every time `execute` runs. A comment about this would be useful, in case someone decides to refactor things later.

gianm commented:
The general idea looks good to me @paul-rogers. Therefore, my comments are all at a level below the general idea. Thanks for the patch.

paul-rogers commented on pom.xml:
```diff
@@ -128,7 +128,7 @@
 
         <!-- Allow the handful of flaky tests with transient failures to pass. -->
-        <surefire.rerunFailingTestsCount>3</surefire.rerunFailingTestsCount>
+        <!-- <surefire.rerunFailingTestsCount>3</surefire.rerunFailingTestsCount> -->
```
Good catch. Yeah, this was for debugging so I'd get "clean" failures when I broke stuff. I'll undo it.

paul-rogers commented on sql/src/main/java/org/apache/druid/sql/SqlLifecycleManager.java:
```diff
@@ -22,40 +22,48 @@
 @LazySingleton
 public class SqlLifecycleManager
 {
+  public interface Cancellable
```
Little did I know I was affecting an English accent all these years. I agree with your analysis. Looks like "cancellation" is the more common form, in both dialects, so kept that.

paul-rogers commented on sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidTypeSystem.java:
```diff
@@ -29,7 +29,8 @@ public class DruidTypeSystem implements RelDataTypeSystem
   /**
-   * Druid uses millisecond precision for timestamps internally. This is also the default at the SQL layer.
+   * Druid uses millisecond precision for timestamps internally. This is also
+   * the default at the SQL layer.
```
Interesting. While the code can sprawl to 120 characters, it is actually quite hard to read comments at that width. Books, magazines and newspapers have limited width for this reason. The old 120-column green-bar paper has long since given way to 80 column (approx.) letter-sized pages.

I wonder, should we allow our code to be 120 characters, but encourage 80 for comments? Here's [a random reference](https://spin.atomicobject.com/2016/09/24/improving-web-typography/#:~:text=Takeaway%3A%20Set%20text%20column%20widths,around%2075%20characters%20helps%20readability.) which suggests I'm not alone in this opinion.

If we want to compress the code, loosening our rules for `{` (put on same line as method declaration) and ')' (append to the last parameter) is a better way to do so.

In any case, reverted the change here.

gianm replied:
> I wonder, should we allow our code to be 120 characters, but encourage 80 for comments? Here's [a random reference](https://spin.atomicobject.com/2016/09/24/improving-web-typography/#:~:text=Takeaway%3A%20Set%20text%20column%20widths,around%2075%20characters%20helps%20readability.) which suggests I'm not alone in this opinion.

Yeah I guess the issue is that shorter lines lead to "taller" code, and displays are generally wider than they are tall.

My feeling is that we need not be authoritarian: the only standard we need across the codebase is that line length for new code should be somewhere between 80 and 120 chars. I wouldn't have said anything here if this was a new comment. I only mentioned it because I didn't think there was a reason to change the pre-existing formatting.

paul-rogers commented on sql/src/main/resources/saffron.properties:
```diff
@@ -0,0 +1,28 @@
+# under the License.
+#-------------------------------------------------------------
+#
+# Properties for Calcite (formerly known as "Saffron").
```
I think it had several names. I see lots of references in Drill code to "Optiq" which seems to have been the name after "Saffron" and before "Calcite." I always thought that "Calcite" used a library called "Saffron", but reading the initialization code revealed that is not actually the case.

Oh, the useless facts with which we clutter our brains...

paul-rogers reviewed (commented):
@gianm, thanks for the review. Revised the code per your comments. And, per your offline comments, revised the PR description to more clearly describe the desire to split the "direct" vs. "prepared" statement (AKA "query") types.

gianm commented on sql/src/main/java/org/apache/druid/sql/AbstractStatement.java:
```diff
@@ -165,12 +180,12 @@ protected PlannerResult plan(DruidPlanner planner)
   public Set<ResourceAction> resources()
   {
-    return queryResource;
+    return cancelationResourceActions;
```
cancellation (it's the only form with two Ls according to the most-specific American English technical style guide I could find: https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/cancel-canceled-canceling-cancellation; and a variety of dictionaries including MW.)

gianm commented on processing/src/main/java/org/apache/druid/query/QueryInterruptedException.java:
```diff
@@ -44,7 +44,7 @@
 {
   public static final String QUERY_INTERRUPTED = "Query interrupted";
-  public static final String QUERY_CANCELLED = "Query cancelled";
+  public static final String QUERY_CANCELED = "Query cancelled";
```
Too bad the "cancelled" spelling is a documented error code. Echoes of the HTTP "Referer" header.

paul-rogers replied:
I've noticed a number of British spellings in the code. Must have been done by a Brit. Perhaps we can credit this spelling to that person as well. Though we can't change the text, it is safe to change the constant name, right? Added a comment to explain the disparity.

gianm replied:
Yeah, it's safe to change the constant name.

gianm reviewed (commented):
Just one spelling-related comment. After that (and the merge conflict) are fixed up this patch LGTM.

gianm commented:
Looks like there is an IT issue due to changes in how exceptions are wrapped.

gianm commented:
Looks like there are still some error-wrapping shenanigans in the same ITs.

paul-rogers commented:
It turns out that JDBC defines an error code for security issues. Modified the Druid driver to catch security errors and map them the Avatica exception which will report the JDBC error. Also modified the tests to use the new "injector builder" classes. Let's see if the tests are happy with these changes.

paul-rogers commented:
@gianm, clean build except for the known-flaky `PrefetchableTextFilesFirehoseFactoryTest`. This can be merged if you want to redo your PR on top, or we can wait for yours, and I'll redo this one on top of yours.

gianm commented:
Yours was clean first 🙂

I'll do the merge in #12897.
